internal/api/external.go

```diff
@@ -68,38 +67,25 @@ func (a *API) ExternalProviderRedirect(w http.ResponseWriter, r *http.Request) e
 	redirectURL := a.getRedirectURLOrReferrer(r, query.Get("redirect_to"))
 	log := observability.GetLogEntry(r)
 	log.WithField("provider", providerType).Info("Redirecting to external provider")
+	if err := validatePKCEParams(codeChallengeMethod, codeChallenge); err != nil {
+		return err
+	}
+	flowType := getFlowFromChallenge(codeChallenge)
 
 	flowStateID := ""
-	switch true {
-	case flowType == PKCE && (codeChallenge == "" || codeChallengeMethodParam == ""):
-		return badRequestError("code challenge and code challenge method are required to perform PKCE")
-	case flowType == PKCE && codeChallenge != "" && codeChallengeMethodParam != "":
-		var codeChallengeMethod models.CodeChallengeMethod
-		switch strings.ToLower(codeChallengeMethodParam) {
-		case "plain":
-			codeChallengeMethod = models.Plain
-		case "s256":
-			codeChallengeMethod = models.SHA256
-		default:
-			return badRequestError("code challenge method is unsupported")
-		}
-		if valid, err := isValidCodeChallenge(codeChallenge); !valid {
+	var codeChallengeMethodType models.CodeChallengeMethod
+	if flowType == models.PKCEFlow {
+		if codeChallengeMethodType, err = models.ParseCodeChallengeMethod(codeChallengeMethod); err != nil {
```

Comment:
```suggestion
        codeChallengeMethodType, err := models.ParseCodeChallengeMethod(codeChallengeMethod)
	if flowType == models.PKCEFlow {
		if err != nil {
```

Reply:
moving the `codeChallengeMethodType` instead the if block because it's not used outside of it

Reply:
Hmm I think `ParseCodeChallenge` can't handle empty values  of codeChallengeMethod though (which happens during implicit flow)